File: Diagrammatize.Design.Pattern/code/cxx/FactoryMethod/A2/idcard/IDCardFactory.cpp
#include "IDCardFactory.hpp"
#include "IDCard.hpp"

Product* IDCardFactory::createProduct(const std::string& owner) {
    return new IDCard(owner, serial++);
}

void IDCardFactory::registerProduct(Product* product) {
    IDCard* card = static_cast<IDCard *>(product);
    database[card->getSerial()] = card->getOwner();
}

const std::unordered_map<int, std::string>& IDCardFactory::getDatabase() {
    return database;
}

File: Diagrammatize.Design.Pattern/code/python/FactoryMethod/A2/idcard/IDCardFactory.py
from framework.Factory import Factory
from .IDCard import IDCard

class IDCardFactory(Factory):
    def __init__(self):
        self.database = dict()
        self.serial = 100

    def createProduct(self, owner):
        card = IDCard(owner, self.serial)
        self.serial += 1
        return card

    def registerProduct(self, product):
        card = product
        self.database[card.getSerial()] = card.getOwner()

    def getDatabase(self):
        return self.database
